*Fix #21807: Implement click journaling service to track user actions  and reproduce server errors that originate from the frontend

mosin74 opened the pull request:
## Overview

1) This PR fixes or fixes part of https://github.com/oppia/oppia/issues/21807.
2) This PR introduces a Click Tracking Service in the frontend to improve error diagnostics. By tracking user interactions (clicks on elements with .e2e-* classes), this feature helps in reproducing frontend errors more easily by providing a history of actions leading to the error. The following changes are implemented:

-Click Tracking Service:
A new service, ClickTrackerService, has been created to track user clicks on elements with .e2e-* classes. It records the last
50 clicks (ensuring memory efficiency) and stores the element's class names in a list.

-Error Handler Integration:
The service has been integrated with the app-error-handler.ts so that when an error occurs, the most recent 50 clicks are
appended to the error report. This will be sent to the backend with the error message, which will help identify the user
actions leading to the error.

-Backend Error Reporting:
The stack trace for the frontend error, along with the string "Last {{numClicks}} clicks: [..., ..., ...]" containing the last 50
recorded clicks, is sent to the backend for better debugging.

-Data Security:
This functionality only triggers when an error occurs. If no error is logged, no click data is sent to the backend to maintain
privacy and security.

-Memory Management:
To avoid excessive memory usage, only the last 50 clicks are stored, and this limit ensures that the log size stays within the
required 16 KB limit.

## Essential Checklist

Please follow the instructions for making a code change.

[x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
[x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
[x] I have written tests for my code.
[x] The PR title starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
[x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct
![Screenshot (64)](https://github.com/user-attachments/assets/f87ed0f7-3519-4ecf-a650-a7ce141bf809)

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

mosin74 committed:
Implemented click tracking service with 50 clicks and maximum size of logs upto 16KB

mosin74 committed:
Correcting linter Checks

mosin74 committed:
Removed unwanted Chrome .deb file

Vir-8 commented:
Hi Mosin, thanks for the PR! Can you link the issue it's fixing, in the PR description

mosin74 commented:
Hi @Vir-8 , thanks for reviewing! I've updated the PR description with the linked issue. Let me know if there's anything else needed.

Vir-8 commented on .gitignore:
```diff
@@ -77,3 +77,4 @@ firebase-debug.log
 debug.log
 # development logs and status entries
 .dev
+google-chrome-stable_current_amd64.deb
```
Hey, why are we adding this here? Where did the deb come from, and is everyone expected to have it when they run Oppia locally now?

mosin74 replied:
It's mistakenly installed in the Oppia directory. I added it to gitignore to ensure it's not unintentionally pushed. chrom is available in the directory., but I need to remove it before pushing. I forgot earlier sorry for the inconvenience. It's now removed.

Vir-8 commented on core/templates/pages/oppia-root/app-error-handler.ts:
```diff
@@ -97,11 +99,22 @@ export class AppErrorHandler extends ErrorHandler {
     }
 
+    // Retrieveing the last recorded clicks.
+
```
Let's remove this blank line

mosin74 replied:
Done with

Vir-8 commented on core/templates/pages/oppia-root/app-error-handler.ts:
```diff
@@ -97,11 +99,22 @@ export class AppErrorHandler extends ErrorHandler {
+      '',
+      formattedClickHistory, // Appended click history here.
+      '',
+      'Stacktrace :' + error?.stack,
```
This stack trace is a bit confusing, can we simplify it and add comments explaining what's being done?

mosin74 replied:
I made some changes. Take a look if you think they're good to go, let me know. Otherwise, please suggest what you think should be done.

Vir-8 commented on core/templates/pages/oppia-root/app.module.ts:
```diff
@@ -54,6 +54,9 @@ import * as hammer from 'hammerjs';
 import {I18nModule} from 'i18n/i18n.module';
 
+// Services .
+import {ClickTrackerService} from 'services/contextual/click-tracker.service';
```
We seem to be importing services under miscellaneous above itself, so let's import this one along with those (instead of a separate block), or we could move the other services here. Either works but let's maintain consistency

mosin74 replied:
Done

Vir-8 reviewed (changes requested):
Hi @mosin74, thanks for the PR! Left a few comments but looking good. Also seems like you haven't maintained frontend coverage (check CI logs), so please write frontend tests for your code.

mosin74 committed:
adding click-tracker.service.spec.ts and commnets for stackTrace in app-error-handler

mosin74 committed:
Removing google-chrome-stable_current_amd64.deb from .gitignore

mosin74 committed:
adding in echo azure-cognitiveservices-speech==1.41.1 >> requirements.in

mosin74 committed:
Fix Linter check

mosin74 committed:
improve error handling in ClickTrackerService tests

mosin74 committed:
tests for  ClickTrackerService

mosin74 commented:
@Vir-8, @U8NWXD I've completed the requested changes. Let me know if there's anything else needed. Thank you!

mosin74 commented:
Dear @Vir-8 and @U8NWXD , The requested changes have been made, and I would appreciate your review at your earliest convenience.

Vir-8 commented:
Hi @mosin74, please do not resolve comments since it makes it difficult for reviewers to keep track, reply to them instead. Also once you've made the changes requested and replied to comments, please assign the reviewers to take a look again. Thanks!

https://github.com/oppia/oppia/wiki/Rules-for-making-PRs#step-4-create-a-pull-request

mosin74 commented:
Thank you so much, @Vir-8! I understand it now. I'm fixing the failed checks, and once that's done, I'll assign the reviewers. It will take some time, but I'm starting to get the hang of the workflow. I really appreciate your help!

mosin74 committed:
Adding comments and test for click-tracker.service

mosin74 committed:
Linter Checks

mosin74 committed:
Linter Checks

mosin74 commented:
@Vir-8 PTAL ,
Before, I was failing two checks: one for the Frontend unit test and one for the Backend unit test. The Backend issue is now resolved—it was related to a duplicate entry for 'azure-cognitiveservices-speech' in the test. 
The failing test was:

**"scripts.install_python_prod_dependencies_test InstallBackendPythonLibsTests.test_uniqueness_of_normalized_lib_names_in_requirements_file."**

For the Frontend unit test, failures occurred in:
**core/templates/pages/oppia-root/oppia-root.component.ts
core/templates/serivces/contextual/click-tracker-service.ts (not fully tested)**

I added an additional test for click-tracker-service.ts, which resolves that issue. However, oppia-root.component.ts is still failing. I don't think it necessarily needs testing since it's only injecting dependencies, but it would be good to add if needed. What do you think? I’ve also added comments in oppia-root.component.ts could you take a look?

The **Full-stack tests / Acceptance** testing is failing, but it was fine before. I don’t think I made any changes that would affect it. I suspect the issue is related to a time limit.

Sorry for the delay was busy with fasting and some important college tasks, so I couldn't focus here. Apologies for that.

mosin74 commented:
@Vir-8 PTAL

mosin74 commented:
@U8NWXD  , @Vir-8 

[Full-stack tests / Acceptance (curriculum-admin/create-edit-and-delete-classroom) (pull_request)](https://github.com/oppia/oppia/actions/runs/13727569207/job/38399359913?pr=21950)

The test is passing successfully locally, but it is failing in the CL environment. Issue is likely related to factors specific to the CL environment. 
Could you please re-run all the CL checks? For reference, here is the local test result showing that the test passes without issues:

![Screenshot 2025-03-11 034956](https://github.com/user-attachments/assets/a14922ca-9372-43c7-b30c-b7396c000524)

seanlip commented:
@mosin74 Please follow these instructions if CI checks fail on your PR. Thanks. https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR

U8NWXD commented on requirements.in:
```diff
@@ -4,6 +4,7 @@
 #
 # to update `requirements.txt` and `third_party/python_libs` accordingly.
 apache-beam[gcp]==2.43.0
+azure-cognitiveservices-speech==1.41.1
```
why are we adding this? I don't see it used anywhere in this PR

mosin74 replied:
While running tests , I encountered an issue related to this . After investigating, I found that this was present in the "develop" branch, so I added it. I look forward to resolving this issue.

U8NWXD reviewed (changes requested):
Have we done a review of the privacy implications of this change? It collects new data about users, but since it doesn't appear to have gone through the usual feature development process (no PRD since there are no user-facing changes), we might have missed the usual privacy discussions. Some initial thoughts:

* I think click events will sometimes be enough to uniquely identify a user. For example, if a user clicks to save changes to their preferences page, then someone could check which users' preferences were updated around the time of the error report to identify the user. Then they'd know the 50 clicks that the user made leading up to the error.
* Users might not expect that their click history is being recorded. It's similar to websites recording keystrokes, which was not well received when publicized a few years back: https://www.wired.com/story/leaky-forms-keyloggers-meta-tiktok-pixel-study/
* A possible remedy might be to give users a choice about whether to submit their click data. When we display an error message, we can give them an option to submit debugging information that includes their recent click history.

/cc @seanlip

mosin74 commented:
Hi @seanlip,  

Right now, we only collect the class names of clicked elements. This way, we gather very little data and make sure users can’t be identified, protecting their privacy.  

If an error happens, the system will ask the user if it’s okay to send their recent click data to help fix the issue. If the user agrees, the data will be sent; if not, nothing will be shared. These changes were requested by @U8NWXD.  

Let me know if I got anything wrong or if you need any changes.

seanlip commented:
@U8NWXD, thanks for pointing this out! I had discussed this with @BenHenning before filing the issue, but revisited it again based on your comment.

- In terms of whether what we plan to do here is OK, we think it is. This is because click events are stored only in the server logs for debugging (and only sent in the case of errors); they are not associated with a user, and we tightly control access to the server logs which are deleted after 30 days. Only team members with datastore access would be able to do the join operation that you describe, and that is a very restricted set of folks. We had also explicitly decided not to record keystrokes because we didn't want to inadvertently have sensitive info recorded in the logs (and that shouldn't be possible with a pattern of clicks).

- However, we do agree that there is a gap in the proposed solution, in terms of whether what we plan to do here is expected. The issue with showing a pop-up, though, is that the original error ends up being in a separate log message from the user-actions information, and this means it won't be easily findable through a search for instances of the error.

After some brainstorming, we came up with the following suggestion:

- When an error arises, then, together with the orange error message that shows up on the page, show a pop-up asking the user whether they're OK with sending us debugging information, and _show the full set of information that will be sent to the server_ so that they can look at it and see that it contains nothing sensitive. (We should probably also explain that this list is only stored locally on their machine and does not sent to us unless they say it's OK.)
- If the user consents, then log another error with the server that includes the additional information.
 
@U8NWXD What do you think about this approach?

Also @mosin74 I'm afraid that this might make this issue a bit harder than originally expected -- we would need to get advice from the design team on the best way to display the error. I'll start a thread for that after getting confirmation from @U8NWXD, but just wanted to let you know that you might need to wait a little bit before doing more work on this issue -- sorry about that.

mosin74 commented:
@seanlip , thanks for breaking this down!  , @U8NWXD I think the idea of showing a pop-up with clear details about what data gets sent and letting users decide is a solid way to balance debugging and transparency.

I’ll hang tight for updates and look forward to hearing more once the design and next steps are figured out. Appreciate you keeping me in the loop!

In the meantime, I’ll work on other stuff. Thanks again for the review!

U8NWXD commented:
@seanlip your proposed approach sounds good to me!

mosin74 commented:
> @seanlip your proposed approach sounds good to me!

@seanlip, @U8NWXD, should I proceed further, or should I wait for more information/clarification ?
Could you please share the link to the open thread for the design team on the best way to display this error? I would be grateful if I could contribute to the design as well. Thank you!

seanlip commented:
@mosin74 Please don't proceed further on this PR. I haven't had the time to file the design issue yet, it's been a rather hectic weekend. I'll probably leave a pointer to it in the original issue thread when I do that.

I'm not sure whether the design team has a plan for developers wanting to do the design mocks, but you can probably ask them about how they'd like to collaborate after I've created the issue.

seanlip commented:
Just to avoid doubt, I'm closing this PR -- we can't work on this until the design team has looked at it, and that may take a while. Filing the issue is still on my to-do list.

mosin74 commented:
@seanlip If you'd like to revisit this, please let me know. I think it would be helpful if you could provide the Figma design, and then I can proceed with it further.

seanlip commented:
Thanks, we'll update the original issue with the requirements once the design is done. Please expect it to take a while, though.
